compiler/rustc_errors/src/json.rs

```diff
@@ -180,6 +181,8 @@ struct Diagnostic {
     children: Vec<Diagnostic>,
     /// The message as rustc would render it.
     rendered: Option<String>,
+    /// Extra tool metadata
+    tool_metadata: ToolMetadata,
 }
```

Comment:
OK @jsgf I took a look at this struct. I think the way to skip emitting the `tool_metadata` field would be to implement the `Encodable` trait *manually* here. It's kind of a pain, for sure, though I guess not *so* bad. You would basically write some code like

```rust
impl<E: Encoder> Encodable<E> for Diagnostic {
    fn encode(&self, s: &mut E) -> Result<(), S::Error> {
        s.emit_struct_field("message", 0, |s| self.message.encode(s))?;
        ...
        if self.tool_metadata.is_some() {
            s.emit_struct_field("message", N, |s| self.message.encode(s))?;
        }
        Ok(())
    }
}
```

Reply:
Obviously a macro could reduce some of the boilerplate here :)

Reply:
I did consider that, but I was concerned that it would become a maintenance liability (add a field without updating the encoder).

Reply:
To avoid this, I usually use a desugaring like

```rust
let Struct { field1, field2 } = value;
execute(field1, field2);
```

this way, if fields are missing, you get a compilation error